### DISCUSSION TITLE
How do we create a plugin.so file with python code for a server in the Shadow TOR simulation ?

### DISCUSSION THREAD

**Discussion by Shubhi-73** [shadow/shadow#3251] [Q&A] [answered]
Function of the plug-in: to execute bash scripts and append the output to logs.

**Comment by jtracey** [comment, marked as answer]
Shadow doesn't need plugins any more, you can just configure processes directly. Check out the [current documentation](https://shadow.github.io/docs/guide/) for some examples on how to do that. Note that executing bash scripts instead of doing everything in a python (or whatever) process will rely heavily on fork/exec. That won't work in Shadow 3.0. It might in the latest git version of Shadow, but it isn't fully implemented (see the [relevant milestone](https://github.com/shadow/shadow/milestone/53) for some hints on what does and doesn't work).

**Reply by robgjansen** [reply]
We have an example of running a python process on a server here:
https://shadow.github.io/docs/guide/getting_started_basic.html#configuring-the-simulation

**Reply by Shubhi-73** [reply]
Thank you for the resources @robgjansen @jtracey, I was able to execute a few scripts. 
I’m planning to run the python code (involves scanning for open ports) directly as a process in the simple TOR network example. I’m repeatedly getting errors. Are there any changes in configurations that are required?
Note- the cURL example with `python3 http.server` is working just fine.

**Shadow.yaml**

```
relay4:
    network_node_id: 0
    processes:
    - path: tor
      args: --Address relay4 --Nickname relay4
            --defaults-torrc torrc-defaults -f torrc
      start_time: 60
      expected_final_state: running
    - path: python3
      args: ~/shadow/examples/docs/tor/test.py
      start_time: 60
      expected_final_state: running
```
**Console**

```
** Starting Shadow v3.0.0-898-g25cb58fc1 2023-12-07--20:00:41 with GLib v2.72.4
00:00:00.687688 [160790:shadow-worker] 00:01:00.000464030 [ERROR] [relay4:11.0.0.8] [process.rs:1527] [shadow_rs::host::process] process 'relay4.python3.10.1001' exited with status Normal(2); expected e
nd state was running but was exited: 2
00:00:05.034148 [160768:shadow] n/a [ERROR] [n/a] [main.rs:403] [shadow_rs::core::main::export] Failed to run the simulation
00:00:05.034335 [160768:shadow] n/a [ERROR] [n/a] [main.rs:403] [shadow_rs::core::main::export] 
00:00:05.034470 [160768:shadow] n/a [ERROR] [n/a] [main.rs:403] [shadow_rs::core::main::export] Caused by:
00:00:05.034533 [160768:shadow] n/a [ERROR] [n/a] [main.rs:403] [shadow_rs::core::main::export]     1 managed processes in unexpected final state
** Shadow did not complete successfully: Failed to run the simulation
```

**Reply by jtracey** [reply]
That means the process exited with an error value (2), you should check the process's logs in the `shadow.data/hosts/relay4/` directory to see what went wrong.

**Reply by robgjansen** [reply]
It looks to me like you told shadow that your python3 process should be running at the end of the simulation (`expected_final_state: running`) but your `test.py` script actually ended before the simulated ended. If you actually intended your test script to end, then you can just change your python3 process to have `expected_final_state: {exited: 0}`.

See here: https://shadow.github.io/docs/guide/shadow_config_spec.html#hostshostnameprocessesexpected_final_state

**Reply by stevenengler** [reply]
Your problem is the line `args: ~/shadow/examples/docs/tor/test.py`. Shadow doesn't do shell expansion (such as `~` to `/home/me/`) of program arguments. Python returns an exit code of 2 when the given file isn't found. Like mentioned above, you should see Python's error message in the stderr log file. And you'll also still want the `expected_final_state` fix.
